materialize/resources/resource_connection_aws_privatelink.go

```diff
@@ -0,0 +1,209 @@
+	connectionType               string
+	privateLinkServiceName       string
+	privateLinkAvailabilityZones []string
+}
+
+func newConnectionAwsPrivatelinkBuilder(connectionName, schemaName, databaseName string) *ConnectionAwsPrivatelinkBuilder {
+	return &ConnectionAwsPrivatelinkBuilder{
+		connectionName: connectionName,
+		schemaName:     schemaName,
+		databaseName:   databaseName,
+	}
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) ConnectionName(connectionName string) *ConnectionAwsPrivatelinkBuilder {
+	b.connectionName = connectionName
+	return b
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) SchemaName(schemaName string) *ConnectionAwsPrivatelinkBuilder {
+	b.schemaName = schemaName
+	return b
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) PrivateLinkServiceName(privateLinkServiceName string) *ConnectionAwsPrivatelinkBuilder {
+	b.privateLinkServiceName = privateLinkServiceName
+	return b
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) PrivateLinkAvailabilityZones(privateLinkAvailabilityZones []string) *ConnectionAwsPrivatelinkBuilder {
+	b.privateLinkAvailabilityZones = privateLinkAvailabilityZones
+	return b
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) Create() string {
+	q := strings.Builder{}
+	q.WriteString(fmt.Sprintf(`CREATE CONNECTION %s.%s.%s TO AWS PRIVATELINK (`, b.databaseName, b.schemaName, b.connectionName))
+
+	q.WriteString(fmt.Sprintf(`SERVICE NAME '%s',`, b.privateLinkServiceName))
+	q.WriteString(fmt.Sprintf(`AVAILABILITY ZONES ('%s')`, strings.Join(b.privateLinkAvailabilityZones, "', '")))
+
+	q.WriteString(`);`)
+	return q.String()
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) ReadId() string {
+	return fmt.Sprintf(`
+		SELECT mz_connections.id
+		FROM mz_connections
+		JOIN mz_schemas
+			ON mz_connections.schema_id = mz_schemas.id
+		JOIN mz_databases
+			ON mz_schemas.database_id = mz_databases.id
+		WHERE mz_connections.name = '%s'
+		AND mz_schemas.name = '%s'
+		AND mz_databases.name = '%s';
+	`, b.connectionName, b.schemaName, b.databaseName)
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) Rename(newConnectionName string) string {
+	return fmt.Sprintf(`ALTER CONNECTION %s.%s.%s RENAME TO %s.%s.%s;`, b.databaseName, b.schemaName, b.connectionName, b.databaseName, b.schemaName, newConnectionName)
+}
+
+func (b *ConnectionAwsPrivatelinkBuilder) Drop() string {
+	return fmt.Sprintf(`DROP CONNECTION %s.%s.%s;`, b.databaseName, b.schemaName, b.connectionName)
+}
+
+func connectionAwsPrivatelinkCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*sqlx.DB)
+
+	connectionName := d.Get("name").(string)
+	schemaName := d.Get("schema_name").(string)
+	databaseName := d.Get("database_name").(string)
+
+	builder := newConnectionAwsPrivatelinkBuilder(connectionName, schemaName, databaseName)
+
+	if v, ok := d.GetOk("service_name"); ok {
+		builder.PrivateLinkServiceName(v.(string))
+	}
+
+	if v, ok := d.GetOk("availability_zones"); ok {
+		azs := v.([]interface{})
```

Comment:
Do you need to initialize a new slice and iterate over the existing one? Would this work?
```
	if v, ok := d.GetOk("availability_zones"); ok {
		builder. PrivateLinkAvailabilityZones(v.([]string))
	}
```

Reply:
Yes, this is exactly how I initially had it, but then realized that it is hitting a nasty panic. eg:

```
panic: interface conversion: interface {} is []interface {}, not []string
```

This is only caught during the tf apply, the go tests don't catch it.
Hence the iteration, any suggestions for improvements would be appreciated!

Reply:
Actually just working on the integration tests for the sources, and noticed the same error occurs for the Postgres Tables as well, eg:

```
| panic: interface conversion: interface {} is map[string]interface {}, not map[string]string
| 
| goroutine 173 [running]:
| terraform-materialize/materialize/resources.sourcePostgresCreate({0x93d928, 0x400010f2c0}, 0x0?, {0x7fa000?, 0x400014b1d0?})
|       /usr/src/app/materialize/resources/resource_source_postgres.go:247 +0x39c
```

Will try to submit a patch for that with the tests from #22

Reply:
Alright. Let's create a separate issue for that for all list types. Thanks